### ISSUE TITLE
Image Viewer - Square patterns are displayed on hi-res images that use toning

### ISSUE THREAD

**Issue by RobbWatershed** [avluis/Hentoid#358]
![Screenshot_2019-07-13-14-48-58-326_me devsaki hentoid](https://user-images.githubusercontent.com/35880555/62428441-5b3c3200-b702-11e9-9276-c00433812cea.png)
![Screenshot_2019-08-05-02-44-29-374_me devsaki hentoid](https://user-images.githubusercontent.com/35880555/62428442-5c6d5f00-b702-11e9-8bee-490c6998821d.png)
![Screenshot_20190713_183001_me devsaki hentoid debug](https://user-images.githubusercontent.com/35880555/62428443-5f684f80-b702-11e9-8057-6c95aa0cb0ce.jpg)

**Comment by RobbWatershed** [comment]
For the record :
- Glide does a slighlty less horrible job (still not satisfying) and is bad at zooming hi-res images
- Using `RapidDecoder` as an alternate image decoder does not improve anything visually

**Comment by RobbWatershed** [comment]
I looked into photo editing software like photoshop and found that there are a wide variety of resampling algorithms that has different qualities
Most efficient but least quality is an algorithm commonly called "nearest neighbor". if an image was scaled down to half it's size then the algorithm just samples every other pixel so it results in a new image that's half the size the original.
This results in artifacts like in that issue ticket but is also very fast
Then there's algorithms like bilinear, trilinear, huffman, lancsoz, etc. More quality at the cost of more processing
So in summary it's a balance
More importantly if we want to improve the subsampling quality we can't do it on-the-fly

**Comment by RobbWatershed** [comment]
------------------------------------

Now how does this relate to the problem we're trying to solve? Well first we don't know how SubsamplingImageView resizes images that causes issue 358. We also don't know if there's any way for us to change it's subsampling strategy.

I'm presuming the best strategy for us is to use nearest neighbor first for quickly showing the image, then decoding the image again using single-pass bilinear to show a better quality. This will read and decode the file twice but single-pass bilinear is the most accurate resizing algorithm readily available in the android sdk.

https://github.com/avluis/Hentoid/issues/358

**Comment by RobbWatershed** [comment]
tweaked the decoder to do what you suggested and used the image vs. screen ratio to determine inDensity and inTargetDensity. Result is the same as before

**Comment by RobbWatershed** [comment]
Useful links
https://stackoverflow.com/questions/39222769/bitmap-scale-destroy-quality

https://stackoverflow.com/questions/2517641/how-to-draw-and-scale-a-bitmap-on-a-canvas-using-bicubic-interpolation-in-androi

**Comment by RobbWatershed** [comment]
Test notes
- Using `Bitmap.createScaledBitmap` gives the same output as `canvas.drawBitmap`

**Comment by RobbWatershed** [comment]
See #510 for a visual comparison of the three resizing methods
